contracts/helper-data/HelperDataFactory.sol

```diff
@@ -0,0 +1,353 @@
+import {Paginator} from "@solarity/solidity-lib/libs/arrays/Paginator.sol";
+import {Vector} from "@solarity/solidity-lib/libs/data-structures/memory/Vector.sol";
+
+import {IHelperDataFactory} from "../interfaces/helper-data/IHelperDataFactory.sol";
+
+contract HelperDataFactory is
+    IHelperDataFactory,
+    OwnableUpgradeable,
+    UUPSUpgradeable,
+    ADeployerGuard
+{
+    using Paginator for *;
+    using EnumerableSet for *;
+    using EnumerableMap for EnumerableMap.UintToAddressMap;
+    using Vector for Vector.AddressVector;
+
+    bytes32 private constant HELPER_DATA_FACTORY_STORAGE_SLOT =
+        keccak256("unforgettable.contract.helper.data.helper.data.factory.storage");
+
+    struct HelperDataFactoryStorage {
+        EnumerableSet.AddressSet helperDataManagers;
+        EnumerableSet.AddressSet registeredAccounts;
+        mapping(address => AccountData) accountsData;
+    }
+
+    modifier onlyHelperDataManager() {
+        _checkHelperDataManager(msg.sender);
+        _;
+    }
+
+    modifier onlyRegisteredAccount(address account_) {
+        _onlyRegisteredAccount(account_);
+        _;
+    }
+
+    constructor() ADeployerGuard(msg.sender) {
+        _disableInitializers();
+    }
+
+    function _getHelperDataFactoryStorage()
+        private
+        pure
+        returns (HelperDataFactoryStorage storage _hdfs)
+    {
+        bytes32 slot_ = HELPER_DATA_FACTORY_STORAGE_SLOT;
+
+        assembly ("memory-safe") {
+            _hdfs.slot := slot_
+        }
+    }
+
+    function initialize(address[] calldata initialManagers_) external initializer onlyDeployer {
+        __Ownable_init(msg.sender);
+
+        _addHelperDataManagers(initialManagers_);
+    }
+
+    /// @inheritdoc IHelperDataFactory
+    function addHelperDataManagers(address[] calldata managersToAdd_) external onlyOwner {
+        _addHelperDataManagers(managersToAdd_);
+    }
+
+    /// @inheritdoc IHelperDataFactory
+    function removeHelperDataManagers(address[] calldata managersToRemove_) external onlyOwner {
+        _removeHelperDataManagers(managersToRemove_);
+    }
+
+    /// @inheritdoc IHelperDataFactory
+    function registerAccount(
+        address account_,
+        uint256 partIndex_,
+        uint64 subscriptionEndTime_,
+        bytes32 metadata_,
+        bytes memory data_
+    ) external onlyHelperDataManager returns (address) {
+        _registerAccount(account_, subscriptionEndTime_, metadata_);
+
+        return _submitHelperDataPart(account_, partIndex_, data_);
+    }
+
+    /// @inheritdoc IHelperDataFactory
+    function updateAccountMetadata(
+        address account_,
+        bytes32 newMetadata_
+    ) external onlyHelperDataManager onlyRegisteredAccount(account_) {
+        _updateAccountMetadata(account_, newMetadata_);
+    }
+
+    /// @inheritdoc IHelperDataFactory
+    function updateAccountSubscriptionEndTime(
+        address account_,
+        uint64 newEndTime_
```

Comment:
Can this timestamp be in the past?

Reply:
Here, we also trust the `helperDataManager` to transfer the subscription's end time correctly. At the same time, this function can be used to set the time to 0, thereby updating the account status to `NONE`. This could serve as a ban or a similar mechanism.